File: frontend/src/utils/format.ts
/**
 * Format date from 8 cell symbols to XX.XX.XX.
 * Indices 2 and 5 (0-based) are always dots; indices 0,1,3,4,6,7 are digits.
 */
export function formatDate(dateCells: string[]): string {
  const pad: string[] = Array(8).fill("");
  dateCells.forEach((c, i) => {
    if (i < 8) pad[i] = (c ?? "").trim();
  });
  const digit = (i: number) => {
    const s = pad[i] ?? "";
    return s === "" || s === "E" ? "" : s;
  };
  return (
    digit(0) +
    digit(1) +
    "." +
    digit(3) +
    digit(4) +
    "." +
    digit(6) +
    digit(7)
  );
}

File: backend/app/utils/format.py
"""Shared formatting utilities."""

# Date uses 8 cells; positions 2 and 5 (0-based) are always dot separators.
DATE_DOT_INDICES = (2, 5)
DATE_DIGIT_INDICES = (0, 1, 3, 4, 6, 7)


def format_date_xx_xx_xx(cells: list[str]) -> str:
    """
    Format date from 8 cell symbols to XX.XX.XX.
    Indices 2 and 5 are always dots; indices 0,1,3,4,6,7 are digits.
    Empty/E/whitespace shown as empty.
    """
    if len(cells) < 8:
        return "".join(c or "" for c in cells)

    def digit(c: str | None) -> str:
        if not c or c.strip() in ("", "E"):
            return ""
        return c.strip()

    return (
        digit(cells[0]) + digit(cells[1]) + "."
        + digit(cells[3]) + digit(cells[4]) + "."
        + digit(cells[6]) + digit(cells[7])
    )
